tests/models/test_dagrun.py

```diff
@@ -552,3 +561,131 @@ def with_all_tasks_removed(dag):
+        # upstream, downstream = dag.tasks
+        #
+        # def run_2nd_dagrun():
+        #     dag.run(
+        #         start_date=timezone.datetime(2016, 1, 2, 0, 0, 0),
+        #         end_date=timezone.datetime(2016, 1, 2, 0, 0, 0),
+        #         executor=MockExecutor()
+        #     )
+        #
+        # # doesn't run if downstream task for previous day has a null state
+        # uti_2 = TI(task=upstream, execution_date=timezone.datetime(2016, 1, 2, 0, 0, 0))
+        # uti_2.previous_ti.set_state(State.SUCCESS)
+        # self.assertEqual(uti_2.previous_ti.state, State.SUCCESS)
+        # TODO: clarify with airflow maintainers -- I don't think this triggers an AirflowException, does it?
+        # self.assertRaises(AirflowException, run_2nd_dagrun)
+        # uti_2.refresh_from_db()
+        # self.assertEqual(uti_2.state, State.SCHEDULED)
+        #
+        # # doesn't run if downstream task for previous day has a failed state
+        # dti_1 = TI(task=downstream, execution_date=timezone.datetime(2016, 1, 1, 0, 0, 0))
+        # dti_1.set_state(State.FAILED)
+        # self.assertRaises(AirflowException, run_2nd_dagrun)
+        # uti_2.refresh_from_db()
+        # self.assertEqual(uti_2.state, State.SCHEDULED)
+        #
+        # # runs if downstream task for previous day has a skipped state
+        # dti_1 = TI(task=downstream, execution_date=timezone.datetime(2016, 1, 1, 0, 0, 0))
+        # dti_1.set_state(State.SKIPPED)
+        # run_2nd_dagrun()
+        # uti_2.refresh_from_db()
+        # self.assertEqual(uti_2.state, State.SUCCESS)
+        #
+        # # runs if downstream task for previous day has a success state
+        # uti_2.set_state(State.NONE)
+        # dti_1.set_state(State.SUCCESS)
+        # run_2nd_dagrun()
+        # uti_2.refresh_from_db()
+        # self.assertEqual(uti_2.state, State.SUCCESS)
+        dag_id = 'test_wait_for_downstream'
+        dag = self.dagbag.get_dag(dag_id)
+        upstream, downstream = dag.tasks
+
+        # For ti.set_state() to work, the DagRun has to exist,
+        # Otherwise ti.previous_ti returns an unpersisted TI
+        self.create_dag_run(dag, execution_date=timezone.datetime(2016, 1, 1, 0, 0, 0))
+        self.create_dag_run(dag, execution_date=timezone.datetime(2016, 1, 2, 0, 0, 0))
+
+        dti1 = TI(task=downstream, execution_date=timezone.datetime(2016, 1, 1, 0, 0, 0))
+        uti2 = TI(task=upstream, execution_date=timezone.datetime(2016, 1, 2, 0, 0, 0))
+        uti2.previous_ti.set_state(State.SUCCESS)
+        self.assertEqual(uti2.previous_ti.state, State.SUCCESS)
+
+        for state in State.task_states:
+            dti1.set_state(state)
+            uti2.set_state(State.QUEUED)
+            uti2.run()
+            if state in (State.SUCCESS, State.SKIPPED):
+                self.assertEqual(uti2.state, State.SUCCESS)
+            else:
+                self.assertNotEqual(uti2.state, State.SUCCESS)
```

Comment:
This test looks very similar to the one you added in test_taskinstance -- why is one in test_dagrun and the other in test_taskinstance?

Reply:
The one in `test_taskinstance.py` is really like a pure unit test. There is no `wait_for_downstream` flag being used anywhere there. It's really just a test on the behavior of `TaskInstance.are_dependents_done()`. 

Meanwhile, the tests in `test_dagrun.py` has a wider scope in which the `wait_for_downstream=True` or `depends_on_past=True` are set when initialising the `Operator`'s. The state of the `previous_ti_downstream` and `previous_ti` are manipulated, `ti` is ran, and then we assert what we expect the state of `ti` should be.